### ISSUE TITLE
Array ring buffer wastes capacity 

### ISSUE THREAD

**Issue by franz1981** [jgroups-extras/jgroups-raft#320]
After eons I wanted to reuse this structure for http 2 stream handling in Netty and..

https://github.com/jgroups-extras/jgroups-raft/blob/a31940e56ade2a8075ebbe2069f8aa611779b3d6/src%2Forg%2Fjgroups%2Fraft%2Futil%2FArrayRingBuffer.java#L269

Doesn't seem to account for gaps 😕 

i.e.
1. Add 3 elements 
2. Remove with index the first 2 inserted: head sequence is still there but we have null out two slots
3. Remove the last and observe that we won't get back anymore the full capacity although all 4 elements are gone 

Remove should keep on checking bubbles of nulls to move the head as much as it can, till the tail, if necessary (or stopping at the first non null).

**Comment by franz1981** [comment]
Nice one, but the issue is valid? I have read it on the phone since today I couldn't use the computer, so I could be very wrong

**Comment by wburns** [comment]
Yeah, I implemented this in our version of the ArrayRingBuffer as I didn't want to deal with the empty space. Check the catchUpHeadPointer method on the PR above.

**Comment by franz1981** [comment]
@wburns 
I have this version I implemented for Netty, simplified: https://gist.github.com/franz1981/6277af990ccfe0a5da8542a75d66ce5e

let me know if it can be of any help
